content/blog/2022/2022-08-02-performance-with-completablefutures.md

```diff
@@ -0,0 +1,249 @@
+
+```
+The operation took 2030 ms
+Categories are: [Category{category='Category_1'}, Category{category='Category_2'}, Category{category='Category_3'}, 
+Category{category='Category_4'}, Category{category='Category_5'}, Category{category='Category_6'}, 
+Category{category='Category_7'}, Category{category='Category_8'}, Category{category='Category_9'}, Category{category='Category_10'}]
+```
+
+We see that the operation took 2030 ms, as predicted. Can we do something to increase even more the performance of our
+application? YES!
+
+## Using CompletableFuture with a custom Executor
+
+Now will refactor our client application to take advantage of CompletableFuture and because we have now ten transactions
+to categorize, we will use an `Executor` with a fixed thread pool of ten Threads.
+
+```java
+public class Demo {
+
+    public static void main(String[] args) {
+        Executor executor = Executors.newFixedThreadPool(10);
+        long start = System.currentTimeMillis();
+        var futureCategories = Stream.of(
+                        new Transaction("1", "description 1"),
+                        new Transaction("2", "description 2"),
+                        new Transaction("3", "description 3"),
+                        new Transaction("4", "description 4"),
+                        new Transaction("5", "description 5"),
+                        new Transaction("6", "description 6"),
+                        new Transaction("7", "description 7"),
+                        new Transaction("8", "description 8"),
+                        new Transaction("9", "description 9"),
+                        new Transaction("10", "description 10")
+                )
+                .map(transaction -> CompletableFuture.supplyAsync(() -> CategorizationService.categorizeTransaction(transaction), executor))
+                .collect(toList());
+
+        var categories = futureCategories.stream()
+                .map(CompletableFuture::join)
+                .collect(toList());
+        long end = System.currentTimeMillis();
+
+        System.out.printf("The operation took %s ms%n", end - start);
+        System.out.println("Categories are: " + categories);
+    }
+}
+```
+
+Our client application is trying to call the categorization service by using the method `supplyAsync` that takes as
+arguments a `Supplier` and an `Executor`. Because now we have ten transactions and a custom Executor that has a pool of
+ten threads, we expect that the operation should take around one second. Indeed, the output confirms the expected
+result :
+
+```
+The operation took 1040 ms
+Categories are: [Category{category='Category_1'}, Category{category='Category_2'}, Category{category='Category_3'}, 
+Category{category='Category_4'}, Category{category='Category_5'}, Category{category='Category_6'}, 
+Category{category='Category_7'}, Category{category='Category_8'}, Category{category='Category_9'}, Category{category='Category_10'}]
+```
```

Comment:
Please add a "Conclusion" section that summarizes the main message of the article

Reply:
done